Add GALAXY_COLLECTIONS_PATH_WARNINGS option.

gotmax23 opened the pull request:
##### SUMMARY
This allows users to disable warnings from `ansible-galaxy collection
install` about `--collections_path` missing from Ansible's configured
collections_paths.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ansible-galaxy

##### ADDITIONAL INFORMATION
In Fedora, we package collections as RPMs and use `ansible-galaxy collection install -p %{buildroot}usr/share/ansible/collections` (`%{buildroot}` is the directory that becomes `/` in the package) to install them. Because the destination directory is not part of the builder's `collections_paths`, `ansible-galaxy` prints out useless warnings.

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1345,15 +1348,17 @@ def _execute_install_collection(
-                            "collections paths '%s'. The installed collection will not be picked up in an Ansible "
-                            "run, unless within a playbook-adjacent collections directory." % (to_text(path), to_text(":".join(collections_path))))
+        if collection_path_warnings:
+            collections_path = C.COLLECTIONS_PATHS
```
Putting this under a conditional makes it undefined under the opposite condition.

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1345,15 +1348,17 @@ def _execute_install_collection(
-                            "run, unless within a playbook-adjacent collections directory." % (to_text(path), to_text(":".join(collections_path))))
+        if collection_path_warnings:
+            collections_path = C.COLLECTIONS_PATHS
+            if len([p for p in collections_path if p.startswith(path)]) == 0:
```
You could extend this condition and avoid having to indent a bunch of code:
```suggestion
            display_path_warning = collection_path_warnings and len([p for p in collections_path if p.startswith(path)])
            if display_path_warning:
```

gotmax23 replied:
Indeed. That's probably better.

bcoca reviewed (dismissed):
We will consider this as a configuration option, but not as a new CLI switch

gotmax23 commented:
> We will consider this as a configuration option, but not as a new CLI switch

Understood. I will look into that. I fixed @webknjaz's comments just for the sake of it.

bcoca commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1364,7 +1364,8 @@ def _execute_install_collection(
         collections_path = C.COLLECTIONS_PATHS
-        if len([p for p in collections_path if p.startswith(path)]) == 0:
+        collections_path_warnings = C.GALAXY_COLLECTIONS_PATH_WARNINGS
+        if collections_path_warnings and len([p for p in collections_path if p.startswith(path)]) == 0:
```
just use the constant directly,  no need to reassign

gotmax23 replied:
I was trying to be consistent with `collections_path`, but I've done as you asked in the latest push.

bcoca commented on lib/ansible/config/base.yml:
```diff
@@ -1367,6 +1367,15 @@ GALAXY_COLLECTION_SKELETON_IGNORE:
+  type: bool
+  env: [{name: ANSIBLE_GALAXY_COLLECTIONS_PATH_WARNINGS}]
+  ini:
+    - {key: collections_path}
```
missing 'section' for ini and version_added for general

gotmax23 replied:
I've fixed `ini:`, but I'm not sure what the value of `version_added` should be.

gotmax23 committed:
Add GALAXY_COLLECTIONS_PATH_WARNING option.

This allows users to disable warnings from `ansible-galaxy collection
install` about `--collections-path` missing from Ansible's configured
collections_paths.

s-hertel commented on lib/ansible/config/base.yml:
```diff
@@ -1367,6 +1367,15 @@ GALAXY_COLLECTION_SKELETON_IGNORE:
   ini:
   - {key: collection_skeleton_ignore, section: galaxy}
   type: list
+GALAXY_COLLECTIONS_PATH_WARNINGS:
```
`version_added` should be 2.16 at this point.

```suggestion
GALAXY_COLLECTIONS_PATH_WARNING:
```

s-hertel commented on lib/ansible/config/base.yml:
```diff
@@ -1367,6 +1367,15 @@ GALAXY_COLLECTION_SKELETON_IGNORE:
+  type: bool
+  env: [{name: ANSIBLE_GALAXY_COLLECTIONS_PATH_WARNINGS}]
+  ini:
+    - {key: collections_path, section: galaxy}
```
```suggestion
    - {key: collections_path_warning, section: galaxy}
```

gotmax23 commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1393,7 +1393,10 @@ def _execute_install_collection(
-        if len([p for p in collections_path if p.startswith(path)]) == 0:
+        if (
+            C.GALAXY_COLLECTIONS_PATH_WARNING
+            and len([p for p in collections_path if p.startswith(path)]) == 0
```
Do you mind if I change this to

```suggestion
            and not [p for p in collections_path if p.startswith(path)]
```
?

s-hertel replied:
~That works for me :+1:~

On the other hand, this isn't the only place we check for an empty list like this, so while I would write it as what you did, just leave it as-is please.

s-hertel commented on lib/ansible/config/base.yml:
```diff
@@ -1366,6 +1366,16 @@ GALAXY_COLLECTION_SKELETON_IGNORE:
   type: list
+GALAXY_COLLECTIONS_PATH_WARNING:
+  name: "ansible-galaxy collection install colections path warnings"
+  descriptions: "whether ``ansible-galaxy collection install`` should warn about ``--collection-path`` missing from configured ``collections_paths``"
```
I *think* you can link to the COLLECTIONS_PATHS config by using :ref:\`collections_paths\` (probably want to verify that with @samccann). \`\`collections_paths\`\` would be represented as code and just look like one of the ini defaults fields, even though there are multiple ways to configure.

gotmax23 replied:
I did a test doc build and confirmed this works as expected:

![image](https://user-images.githubusercontent.com/9920591/233459046-2baccf37-5605-495b-8c66-506f7f2968b9.png)

samccann replied:
lol cool cuz I'd have to do the same thing on this one :-)

s-hertel replied:
thanks for checking that!

s-hertel commented on changelogs/fragments/78487-galaxy-collections-path-warnings.yml:
```diff
@@ -0,0 +1,4 @@
+---
+minor_changes:
+  - "Add ``GALAXY_COLLECTIONS_PATH_WARNING`` option to enable/disable
+    ``ansbile-galaxy collection install`` collections_paths warnings."
```
```suggestion
- >-
  Add ``GALAXY_COLLECTIONS_PATH_WARNING`` option to disable the warning
  given by ``ansible-galaxy collection install`` when installing a collection
  to a path that isn't in the configured collection paths.
 ```

s-hertel commented on lib/ansible/config/base.yml:
```diff
@@ -1366,6 +1366,16 @@ GALAXY_COLLECTION_SKELETON_IGNORE:
+  ini:
+    - {key: collections_path_warning, section: galaxy}
+  version_added: "2.16"
+
```
```suggestion

```

s-hertel commented on lib/ansible/config/base.yml:
```diff
@@ -1366,6 +1366,16 @@ GALAXY_COLLECTION_SKELETON_IGNORE:
   type: list
+GALAXY_COLLECTIONS_PATH_WARNING:
+  name: "ansible-galaxy collection install colections path warnings"
+  description: "whether ``ansible-galaxy collection install`` should warn about ``--collection-path`` missing from configured :ref:`collections_paths`"
```
```suggestion
  description: "whether ``ansible-galaxy collection install`` should warn about ``--collections-path`` missing from configured :ref:`collections_paths`"
```

gotmax23 committed:
Apply grammar and formatting fixes

Co-authored-by: Sloane Hertel <19572925+s-hertel@users.noreply.github.com>

s-hertel commented:
Thanks @gotmax23 for the quick updates! Sorry for letting this slip into the backlog.
